src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -119,44 +119,55 @@ static ProjectCodeRefactoringProvider.ExtensionInfo GetExtensionInfo(ExportCodeR
         using (TelemetryLogging.LogBlockTimeAggregated(FunctionId.CodeRefactoring_Summary, $"Pri{priority.GetPriorityInt()}"))
         using (Logger.LogBlock(FunctionId.Refactoring_CodeRefactoringService_GetRefactoringsAsync, cancellationToken))
         {
-            var extensionManager = document.Project.Solution.Services.GetRequiredService<IExtensionManager>();
-            using var _ = ArrayBuilder<Task<CodeRefactoring?>>.GetInstance(out var tasks);
-
-            foreach (var provider in GetProviders(document))
-            {
-                if (priority != null && priority != provider.RequestPriority)
-                    continue;
-
-                tasks.Add(Task.Run(async () =>
+            using var _ = ArrayBuilder<CodeRefactoring>.GetInstance(out var codeRefactorings);
+
+            var providers = GetProviders(document).Where(p => priority == null || p.RequestPriority == priority);
+
+            await ProducerConsumer<CodeRefactoring>.RunAsync(
+                ProducerConsumerOptions.SingleReaderOptions,
+                produceItems: static (callback, args) =>
+                    // Run all providers in parallel to get the set of refactorings for this document.
+                    RoslynParallel.ForEachAsync(
+                        args.providers,
+                        args.cancellationToken,
+                        async (provider, cancellationToken) =>
+                        {
+                            // Log an individual telemetry event for slow code refactoring computations to
+                            // allow targeted trace notifications for further investigation. 500 ms seemed like
+                            // a good value so as to not be too noisy, but if fired, indicates a potential
+                            // area requiring investigation.
+                            const int CodeRefactoringTelemetryDelay = 500;
+
+                            var providerName = provider.GetType().Name;
+                            args.@this.RefactoringToMetadataMap.TryGetValue(provider, out var providerMetadata);
+
+                            var logMessage = KeyValueLogMessage.Create(m =>
+                            {
+                                m[TelemetryLogging.KeyName] = providerName;
+                                m[TelemetryLogging.KeyLanguageName] = args.document.Project.Language;
+                            });
+
+                            using (args.addOperationScope(providerName))
+                            using (RoslynEventSource.LogInformationalBlock(FunctionId.Refactoring_CodeRefactoringService_GetRefactoringsAsync, providerName, cancellationToken))
+                            using (TelemetryLogging.LogBlockTime(FunctionId.CodeRefactoring_Delay, logMessage, CodeRefactoringTelemetryDelay))
+                            {
+                                var extensionManager = args.document.Project.Solution.Services.GetRequiredService<IExtensionManager>();
```

Comment:
>var extensionManager [](http://example.com/codeflow?start=32&length=20)

out of curiosity, why was this moved inside the producer code? Couldn't it have been passed in like it was before?

Reply:
very true.  it was more annoyance around passing the args along :)